### ISSUE TITLE
emb

### ISSUE THREAD

**Issue by SkBlaz** [SkBlaz/py3plex#1250]
You are GitHub Copilot working inside the py3plex repository. Implement “first-class embeddings” as a core ML primitive integrated into DSL v2 builder syntax, with provenance, caching, and pluggable backends (scipy baseline + optional JAX). Do not add new top-level documentation files. Prefer small, well-contained modules and minimal API surface. Maintain backward compatibility.

GOALS

Add .embed(...) as a first-class operation to DSL v2 (Q.nodes() and optionally Q.edges()).

Provide a production-grade node embedding implementation: NetMF.

Provide link embedding as a first-class derivation from node embeddings via operators (hadamard, concat, l1, l2, dot, cosine).

Integrate provenance metadata and performance timings into QueryResult.meta["provenance"] (and dict export).

Add caching keyed by (network_fingerprint, query_ast_hash, embed_config_hash) with explicit invalidation on network mutation/version (if available).

Provide export utilities: QueryResult.embeddings plus .to_numpy("embedding") / .to_pandas(expand_embeddings=...) behavior.

Keep API consistent with existing DSL v2 / AST compilation model: builder → AST → executor → QueryResult.

NON-GOALS

Do not implement deep learning GNNs.

Do not require GPU; JAX is optional.

Do not break existing DSL v2 queries or legacy DSL.

Do not over-engineer new frameworks; integrate with existing patterns (pipeline steps, profiling utilities, provenance).

DELIVERABLES (code changes)

1. DSL v2 Builder API changes

Add embed() to Q.nodes() chain:

Signature: .embed(method="netmf", dim=128, multilayer="supra", layers=None, window=10, negative=1.0, approx="randomized_svd", seed=None, backend="auto", cache=True, cache_namespace="dsl_embeddings")

.embed() must be composable before/after .where(), .from_layers(), .limit(), .order_by() etc.

Store embedding request in AST (new AST node or extended Compute node).

Add .embed_from_nodes(node_embedding_ref, op="hadamard") for edges queries OR allow .embed(op=..., from_nodes=True); pick the cleanest minimal approach.

2. AST + Executor integration

Extend unified AST schema to represent embedding computation:

EmbeddingSpec: {target: "nodes"|"edges", method, dim, multilayer_mode, params, backend, seed, cache, ref_name}

For edges link-features, define LinkEmbedSpec: {source_embedding_ref, op}

Ensure deterministic hashing for AST includes embedding spec so provenance ast_hash changes.

Modify DSL v2 executor pipeline stages:

Stage order should be: bind params → get items (nodes/edges) → filter layers → filter where → compute measures → embed (or embed before compute if required) → ordering/limit → finalize.

Make embedding execution depend on the filtered node set by default (i.e., embed on the subgraph induced by selected nodes in selected layers).

For “multilayer” operators, implement at least:

"union": union graph across chosen layers

"supra": supra-adjacency with interlayer coupling gamma (default 1.0) between same node across layers (configurable gamma)

"per_layer": compute per-layer embeddings then align/fuse (optional; if too much, stub with NotImplementedError and clear message, but implement union + supra fully)

"rw_coupled": optional later; do not implement unless easy.

3. Core Embedding module

Create py3plex/embeddings/ package (or equivalent existing module path) with:

base.py: EmbeddingResult, EmbeddingBackend protocol, Embedder base

netmf.py: NetMF implementation

link_ops.py: link feature operators

cache.py: caching utilities (in-memory LRU + optional disk cache in .py3plex_cache/ if project already has patterns; otherwise in-memory only but keyed robustly)

NetMF implementation requirements:

Input can be NetworkX graph (from network.core_network) or a scipy sparse adjacency built from it.

Provide adjacency construction for multilayer union and supra modes:

union: collapse layers into single graph over unique node IDs (option: treat node+layer as distinct; choose consistent with py3plex node identity tuples; for union across layers, map (id, layer) → id if same id; document in code comments)

supra: treat each (id, layer) as distinct node; add interlayer edges between same id across layers with weight gamma; preserve intralayer edges.

Implement PMI matrix approximation consistent with NetMF:

Use truncated SVD/eig on a transformed matrix. If full PMI is too heavy, implement scalable approximation:

Use randomized_svd on sparse/dense as appropriate.

Provide approx="eigsh" alternative.

Ensure stable randomness control via seed.

Output embeddings:

For nodes query: return embeddings aligned to QueryResult.items order.

Store embedding matrix X shape (n_items, dim) in QueryResult.embeddings["node"].

Also compute embedding_norm per item as a computed attribute for easy ordering/filtering.

4. Backends (scipy baseline + optional jax)

Implement backend="auto"|"scipy"|"jax".

If backend="auto": use jax if installed else scipy.

scipy backend must work without jax.

jax backend:

Use JAX for compute-heavy steps when feasible (e.g., factorization or matmul).

If JAX sparse support is insufficient, fall back to scipy for sparse preprocessing and use jax for dense factorization, with clear code path.

Must not crash if no accelerator; should run on CPU.

Add minimal dependency handling: import guarded; clear error messages.

5. QueryResult extensions

Add QueryResult.embeddings: dict[str, Any] (e.g., {"node": np.ndarray, "edge": np.ndarray}).

Add .to_numpy(kind="embedding") returning embedding matrix aligned to items.

Extend .to_pandas(expand_embeddings=False, embedding_prefix="emb_"):

If expand_embeddings=True, add columns emb_0..emb_{dim-1}.

Otherwise include a single column embedding holding list/np arrays OR omit entirely (choose consistent behavior).

Ensure to_dict() exports embeddings in a safe way (e.g., list of lists) behind a flag include_embeddings=True default False to avoid huge dicts; but keep embeddings accessible via QueryResult in memory.

6. Provenance + profiling integration

Extend provenance schema to include:

provenance["embedding"]: method, dim, multilayer_mode, gamma, window, negative, approx, backend, seed, cache_hit(bool), embed_config_hash

provenance["performance"] timings include embedding stage: embed_ms

Use existing profiling utilities (timed_section, monitor) if present; else lightweight timing with time.perf_counter.

Ensure provenance network_fingerprint captures layers used for embedding and mode.

7. Caching

Implement caching at embedding stage:

Key: (network_fingerprint_hash, ast_hash, embed_config_hash) where embed_config_hash excludes seed if seed is None? Decide: include seed always for correctness.

Store: embedding matrix + mapping of item ids used, so alignment can be validated.

If network has mutation counter (network_version), include it.

Add cache=True/False in embed spec; default True.

On cache hit, set provenance cache_hit=True and skip compute.

8. Tests (must add)

Add unit tests validating:

.embed(method="netmf") returns embeddings with correct shape for nodes queries.

Determinism with fixed seed (same results across runs within tolerance).

Different multilayer modes yield different embeddings (union vs supra) on small toy multilayer.

Link ops produce expected shapes for edges and simple operator correctness (hadamard etc.).

Provenance contains embedding block + performance embed_ms key.

Cache hit path works (second run hits cache).

Keep tests fast (<2s). Use tiny graphs.

9. CLI / Pipeline optional integration (small)

Add a pipeline step ComputeEmbedding (optional) that wraps embedder for non-DSL users.

Do not expand CLI unless trivial; if added, keep it minimal: py3plex embedding --method netmf ... only if already fits command patterns.

IMPLEMENTATION STEPS (follow in order) A) Locate DSL v2 builder classes (Q, QueryBuilder, QueryResult, executor). Identify where .compute() is represented and executed. B) Add EmbeddingSpec to AST; update hashing and AST summary. C) Add builder .embed() to append embedding spec. D) Implement embedding execution in executor with caching + provenance. E) Implement embeddings module with NetMF + link ops + backend selection. F) Extend QueryResult to carry embeddings and export utilities. G) Add tests.

QUALITY BAR

Type hints everywhere in new code.

Clear exceptions: EmbeddingError (subclass of existing Py3plex exceptions if available).

Minimal public API, but stable and consistent.

No heavy new dependencies required for default path.

No “TODO” placeholders; either implement or raise explicit NotImplementedError with message.

CODE STYLE

Match existing repository conventions.

Keep functions small; avoid giant files.

Add docstrings to public functions/classes.

Prefer numpy/scipy for baseline; JAX optional.

NOW IMPLEMENT.
